R/species.r

```diff
@@ -0,0 +1,114 @@
+#' Infer Species
+#'
+#' @param sset a \code{SigSet}
+#' @param df_as a data.frame of alignment score for each probe.
+#' @param infer.human whether to use the fraction of negative probes to infer human.
+#' @param topN Top n positive and negative probes used to infer species.
+#' @return a list of auc, pvalue, species (NCBI official species names) and taxid.
+#' We infer species based on probes pvalues and alignment score.
+#' AUC was calculated for each specie, y_true is 1 or 0 
+#' for pval < 0.01 or pval > 0.2, respeceively,
+#' y_pred is normalized alignment score (AS / 50).
+#'
+#' Our function works on a single sample.
+#' @importFrom PRROC roc.curve
+#' @import sesameData
+#' @examples
+#' sset <- sesameDataGet('EPIC.1.LNCaP')$sset
+#' df_as <- sesameDataGet() #TODO
+#' inferSpecies(sset)
+#' 
+#' @examples
+#' library(tidyverse)
+#' library(sesame)
+#' library(parallel)
+#' library(readxl)
+#' indir="/mnt/isilon/zhou_lab/projects/20191212_GEO_datasets/IDATs/Mammal40/"
+#' AS_dir="/mnt/isilon/zhoulab/labprojects/20200228_Mouse_Array_Project/20210324_SeSAme_Wubin/data/AS_A_probe_species_matrix.txt.gz"
+#' samplesheet="/mnt/isilon/zhoulab/labsamplesheets/2021/20210325_Mammal40_Merged_GEO_Steve_Wubin_samplesheets.xlsx"
+#' 
+#' sample_info=read_excel(samplesheet)
+#' df_as=read.table(AS_dir,header=T,sep='\t')
+#' rownames(df_as) <- do.call(paste,c(df_as[c("Species","taxid")],sep="|"))
+#' df_as <- t(df_as[,!names(df_as) %in% c("Species","taxid")])
+#' df_as[1:3,1:3]
+#' 
+#' sample_info$IDs <- do.call(paste,c(sample_info[c("NCBI_scientific_name","taxid")],sep="|"))
+#' overlapped_species=intersect(unique(sample_info$IDs),colnames(df_as))
+#' sample_info=sample_info[sample_info$IDs %in% overlapped_species,]
+#' df_as=df_as[,overlapped_species]
+#' df_as=df_as / 50
+#' idats=searchIDATprefixes(indir)
+#' idats=idats[sample_info$Basename]
+#' idats=idats[! is.na(idats)]
+#' 
+#' mft <- readRDS('/mnt/isilon/zhoulab/labprojects/20201124_Horvath_mammal_array/mammal_array_manifest_original.rds')
+#' sset=readIDATpair(idats[1],manifest=mft, platform='mammal')
+#' res=inferSpecies(sset,df_as)
+#' res=inferSpecies(sset,df_as,ret.max=F)
+#' result=do.call(cbind, mclapply(idats, function(x) {
+#' 		sset=readIDATpair(x,manifest=mft, platform='mammal')
+#'      res=inferSpecies(sset,df_as,ret.max=T)
+#'    },mc.cores=8)
+#'    )
+
+#' result=do.call(cbind, mclapply(idats, function(x) {
+#'       sset=readIDATpair(x,manifest=mft, platform='mammal')
+#'       res=inferSpecies(sset,df_as,ret.max=F)
+#'       return(res[,'auc'])
+#'    },mc.cores=8)
+#'    )
+#'
+#' write.table(as.data.frame(result),file="AUC_for_species.txt",sep='\t',quote=FALSE,col.names=T)
+#' @export
+
+inferSpecies <- function(sset,df_as=NULL,infer.human=T,topN=3000,ret.max=T) {
+    pvalue=pval(sset)
+    pvalue <- pvalue[intersect(names(pvalue),rownames(df_as))]
+    pos_probes=sort(pvalue[pvalue< 0.01],decreasing=F)
+    neg_probes=sort(pvalue[pvalue > 0.2],decreasing=T)
```

Comment:
we should make 0.01 and 0.2 arguments of the function

Reply:
threshold_pos and threshold_neg have been added.